### PR TITLE
Tests for users internal API

### PR DESCRIPTION
**Pull request by Eiriknil**

### CODE REVIEW

File: tests/users_internal/users_internal.feature [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        * def postSuccessPayload = read('../../test_files/nva_users_internal/post_success_payload.json')
+        * def postResponse = read('../../test_files/nva_users_internal/post_response.json')
+        * def existingUserPayload = read('../../test_files/nva_users_internal/post_existing_user_payload.json')
+        * def missingTypePayload = read('../../test_files/nva_users_internal/post_missing_type_payload.json')
+        * def wrongTypePayload = read('../../test_files/nva_users_internal/post_wrong_type_payload.json')
+        * def wrongRoleTypePayload = read('../../test_files/nva_users_internal/post_wrong_role_type_payload.json')
+        * def putSuccessPayload = read('../../test_files/nva_users_internal/put_success_payload.json')
+        * def getUser = 'user-internal-get@test.no'
+        * def putUser = 'user-internal-put@test.no'
+        * def notExistingUser = 'not-existing-user'
```
Comment (brinxmat) [inline tests/users_internal/users_internal.feature]: I guess this is non-existent-user as 'not-existing-user' implies that the usr is going to be created

File: tests/users_internal/users_internal.feature [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        Given path getUser
+        When method get
+        Then status 200
+        And match response == getResponse
+
+    Scenario: GET user not found
+        Given path notExistingUser
+        When method get
+        Then status 404
+        And match response.title == 'Not Found'
```
Comment (brinxmat) [inline tests/users_internal/users_internal.feature:29]: Assert that the response contains the status too in all cases, which it should

File: tests/users_internal/users_internal.feature [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        And match response.detail == 'Could not find user with username: ' + notExistingUser
+
+    Scenario: POST user success
+        Given path '/'
+        And request postSuccessPayload
+        When method post
+        Then status 200
+        And match response == postResponse
+
+    Scenario: POST user allready exist
```
Comment (brinxmat) [inline tests/users_internal/users_internal.feature]: Already

File: tests/users_internal/users_internal.feature [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        And match response.title == 'Bad Request'
+        And match response.detail == 'JSON object is missing a type attribute'
+    
+    Scenario: POST wrong type attribute
+        Given path '/'
+        And request wrongTypePayload
+        When method POST
+        Then status 400
+        And match response.title == 'Bad Request'
+        And match response.detail == 'JSON object is missing a type attribute'
```
Comment (brinxmat) [inline tests/users_internal/users_internal.feature:64]: missing, really?

File: tests/users_internal/users_internal.feature [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    
+    Scenario: POST wrong role type attribute
```
Comment (brinxmat) [inline tests/users_internal/users_internal.feature]: This repeats the previous scenario

File: tests/users_internal/users_internal.feature [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        And match response.title == 'Bad Request'
+        And match response.detail == 'JSON object is missing a type attribute'
+    
+    Scenario: PUT success
+        Given path putUser
+        And method GET
+        Then status 200
+        And match response.givenName == 'User internal PUT'
+
+        Given path putUser
```
Comment (brinxmat) [inline tests/users_internal/users_internal.feature]: Missing scenario description on the two following scenarios

File: tests/users_internal/users_internal.feature [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        Given path '/'
+        And request wrongRoleTypePayload
+        When method POST
+        Then status 400
+        And match response.title == 'Bad Request'
+        And match response.detail == 'JSON object is missing a type attribute'
+    
+    Scenario: PUT success
+        Given path putUser
+        And method GET
```
Comment (brinxmat) [inline tests/users_internal/users_internal.feature]: And or When? Following pattern above…

File: tests/users_internal/users_internal.feature (hunk 6 above)
Comment (brinxmat) [reply tests/users_internal/users_internal.feature]: I guess we're updating the same resource here, but the scenarios become dependent on each other, which we don't want

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: tests/users_internal/users_internal.feature (hunk 4 above)
Comment (Eiriknil) [reply tests/users_internal/users_internal.feature:64]: That's what the status says

File: tests/users_internal/users_internal.feature (hunk 5 above)
Comment (Eiriknil) [reply tests/users_internal/users_internal.feature]: Different type element, first one is for the User, this is for the Role

File: tests/users_internal/users_internal.feature (hunk 6 above)
Comment (Eiriknil) [reply tests/users_internal/users_internal.feature]: Technically the same Scenario, testing to see that the User has been updated correctly. I guess I should just add these to the Then part of the Scenario

File: tests/users_internal/users_internal.feature (hunk 4 above)
Comment (brinxmat) [reply tests/users_internal/users_internal.feature:64]: That sounds like a bug to me :D

**Review by brinxmat — DISMISSED** [review]

File: tests/users_internal/users_internal.feature [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        * def wrongTypePayload = read('../../test_files/nva_users_internal/post_wrong_type_payload.json')
+        * def wrongRoleTypePayload = read('../../test_files/nva_users_internal/post_wrong_role_type_payload.json')
+        * def putSuccessPayload = read('../../test_files/nva_users_internal/put_success_payload.json')
+        * def getUser = 'user-internal-get@test.no'
+        * def putUser = 'user-internal-put@test.no'
+        * def nonExistingUser = 'non-existing-user'
+
+    Given url 'https://api.dev.nva.aws.unit.no/users-roles-internal/service/users/'
+
+    Scenario: GET user success
```
Comment (axthosarouris) [inline tests/users_internal/users_internal.feature]: Scenario descriptions should be more detailed:  `GET returns user-details and status OK when requesting existing user and client is authorized'

File: tests/users_internal/users_internal.feature [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        Given path getUser
+        When method get
+        Then status 200
+        And match response == getResponse
+
+    Scenario: GET user not found
```
Comment (axthosarouris) [inline tests/users_internal/users_internal.feature]: Get returns notFound when client requests non-exisiting user'

**Review by axthosarouris — CHANGES_REQUESTED** [review]
I think the scenario titles are too short. I think that the scenario title should describe the expectation of the test, so that a) The reader does not have to read the scenario to understand the fundamental expectations of the test, and b) errors in the test code can be easily traced. E.g. copy-pasting the wrong error code. I am not necessarily requesting for changes, but I think this issue should be resolved with one or another way before we merge the code.

By "resolve" I mean that we either agree that either the title should explain what is being tested (e.g. GET user) or the title should describe shortly what is the intended functionality (e.g. 'GET user returns Unauthorized when client is not authorized'). Personally, I like the second alternative but I am open to all alternatives.

File: tests/users_internal/users_internal.feature (hunk 8 above)
Comment (Eiriknil) [reply tests/users_internal/users_internal.feature]: Agree, I will update descriptions to be more detailed.

**Review by axthosarouris — APPROVED** [review]

**Review by janbj — APPROVED** [review]

**Pull request merged**